### PR TITLE
Gracefully offline and close clients when app shutdown

### LINKED ISSUE #1988: v1.26.0 spamming with "OnDisconnectedAsync" error
With a recent version update, the library started spamming with the given error. It happens whenever application is stopped/restarted.

Error when dispatching '"OnDisconnectedAsync"' on hub.
System.ObjectDisposedException: Instances cannot be resolved and nested lifetimes cannot be created from this LifetimeScope as it (or one of its parent scopes) has already been disposed.
   at Autofac.Core.Lifetime.LifetimeScope.ThrowDisposedException()
   at Autofac.Core.Lifetime.LifetimeScope.BeginLifetimeScope(Object tag)
   at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.OnDisconnectedAsync(HubConnectionContext connection, Exception exception)
   at Microsoft.AspNetCore.SignalR.HubConnectionHandler`1.HubOnDisconnectedAsync(HubConnectionContext connection, Exception exception)
   at Microsoft.AspNetCore.SignalR.HubConnectionHandler`1.RunHubAsync(HubConnectionContext connection)
   at Microsoft.AspNetCore.SignalR.HubConnectionHandler`1.OnConnectedAsync(ConnectionContext connection)
   at Microsoft.AspNetCore.SignalR.HubConnectionHandler`1.OnConnectedAsync(ConnectionContext connection)
   at Microsoft.Azure.SignalR.ServiceConnection.ProcessApplicationTaskAsyncCore(ClientConnectionContext connection)

### PR DESCRIPTION
**Pull request by vicancy**
Fix https://github.com/Azure/azure-signalr/issues/1988 and https://github.com/Azure/azure-signalr/issues/1777

Use graceful shutdown to completely fix the disposable exception issue.

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Constants.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ internal static class Constants
     public const string AsrsDefaultScope = "https://signalr.azure.com/.default";
 
 
-    public const int DefaultCloseTimeoutMilliseconds = 30000;
+    public const int DefaultCloseTimeoutMilliseconds = 5000;
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Common/Constants.cs]: This is claimed in the [GracefulShutdownOptions](https://github.com/Azure/azure-signalr/blob/dev/src/Microsoft.Azure.SignalR/GracefulShutdownOptions.cs#L21C67-L21C89), also need an update?

File: src/Microsoft.Azure.SignalR/HubHost/ServiceHubDispatcher.cs [hunk 2]
```diff
@@ -158,8 +153,17 @@ private async Task GracefulShutdownAsync(GracefulShutdownOptions options, Cancel
         Log.SettingServerOffline(_logger, _hubName);
         await _serviceConnectionManager.OfflineAsync(options.Mode, cancellationToken);
 
-        Log.TriggeringShutdownHooks(_logger, _hubName);
-        await options.OnShutdown(Context);
+        if (options.Mode == GracefulShutdownMode.Off)
+        {
+            // By default we directly close all the client connections
+            Log.CloseClientConnections(_logger, _hubName);
+            await _serviceConnectionManager.CloseClientConnections(cancellationToken);
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR/HubHost/ServiceHubDispatcher.cs:160]: `CloseClientConnections` seems to be called in multiple places, are they doing same tasks?

Comment (vicancy) [reply src/Microsoft.Azure.SignalR/HubHost/ServiceHubDispatcher.cs:160]: I don't think it is called in multiple places, it's passing down from manager to container and finally to the service connection

File: src/Microsoft.Azure.SignalR.Common/Constants.cs (hunk 1 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Common/Constants.cs]: this one is for a client application task timeout, not related to graceful shutdown

**Review by JialinXin — APPROVED** [review]

File: src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs [hunk 3]
```diff
@@ -111,9 +111,29 @@ protected override Task DisposeConnection(ConnectionContext connection)
         return _connectionFactory.DisposeAsync(connection);
     }
 
-    public override Task CloseClientConnections(CancellationToken token)
+    public override async Task CloseClientConnections(CancellationToken token)
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs:114]: Will this similar cleanup interrupt with below function `CleanupClientConnections`?

**Comment by bodamgaard** [comment]
@vicancy will this also fix this issue? :-) #1777

File: src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs (hunk 3 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR/ServerConnections/ServiceConnection.cs:114]: no, close client connections only do close.

**Comment by vicancy** [comment]
> @vicancy will this also fix this issue? :-) #1777

Yes it contains the fix now.

**Review by JialinXin — APPROVED** [review]

**Pull request merged**